src/plotman/graph.py

```diff
@@ -0,0 +1,263 @@
+                # Bucket 267 uniform sort. Ram: 0.920GiB, u_sort min: 0.688GiB, qs min: 0.172GiB.
+                #   or
+                # ....?....
+                #   or
+                # Bucket 511 QS. Ram: 0.920GiB, u_sort min: 0.375GiB, qs min: 0.094GiB. force_qs: 1
+                m = re.search(r'Bucket \d+ ([^\.]+)\..*', line)
+                if m and not 'force_qs' in line:
+                    sorter = m.group(1)
+                    n_sorts += 1
+                    if sorter == 'uniform sort':
+                        n_uniform += 1
+                    elif sorter == 'QS':
+                        pass
+                    else:
+                        print ('Warning: unrecognized sort ' + sorter)
+
+                # Job completion.  Record total time in sliced data store.
+                # Sample log line:
+                # Total time = 49487.1 seconds. CPU (97.26%) Wed Sep 30 01:22:10 2020
+                m = re.search(r'^Total time = (\d+.\d+) seconds.', line)
+                if m:
+                    time_taken = float(m.group(1))
+                    data.setdefault(sl, {}).setdefault('total time', []).append(time_taken)
+                    for phase in ['1', '2', '3', '4']:
+                        data.setdefault(sl, {}).setdefault('phase ' + phase, []).append(phase_time[phase])
+                    data.setdefault(sl, {}).setdefault('%usort', []).append(100 * n_uniform // n_sorts)
+
+                    time_ended = time.mktime(datetime.datetime.strptime(line.split(')')[-1][1:-1], '%a %b %d %H:%M:%S %Y').timetuple())
+                    data.setdefault(sl, {}).setdefault('time ended', []).append(time_ended)
+                    data.setdefault(sl, {}).setdefault('time started', []).append(time_ended - time_taken)
+
+    # Prepare report
+    for sl in data.keys():
+
+        # This array will hold start and end data (in hours)
+        data_started_ended = np.array([[ts, te, te-ts] for
+            ts, te in zip(data[sl]['time started'], data[sl]['time ended'])
+        ]) / (60 * 60)
+
+        # Sift the data so that it starts at zero
+        data_started_ended -= np.min(data_started_ended[:, 0])
+
+        # Sort the rows by start time
+        data_started_ended = data_started_ended[np.argsort(data_started_ended[:, 0])]
+
+        # Create figure
+        num_plots = 4
+        f, _ = plt.subplots(2,1, figsize=(8, 12))
+        ax = plt.subplot(num_plots,1,1)
+        ax.set_title('Plot performance summary')
+
+        create_ax_dumbbell(ax, data_started_ended)
+
+        ax = plt.subplot(num_plots,1,2)
+        create_ax_plotrate(ax, data_started_ended, end=True, window=3)
+
+        ax = plt.subplot(num_plots,1,3)
+        create_ax_plottime(ax, data_started_ended, window=3)
+
+        ax = plt.subplot(num_plots,1,4)
+        create_ax_plotcumulative(ax, data_started_ended)
+
+        ax.set_xlabel('Time (hours)')
+        f.savefig(figfile)
+
+
+if __name__ == '__main__':
```

Comment:
Is it particularly needed for this to be directly runnable rather than just using the plotman CLI access?